src/ao_rocksdb_store.erl

```diff
@@ -0,0 +1,493 @@
+    end;
+resolve(_Opts, Key) when is_binary(Key) ->
+    case item_path(Key) of
+        not_found ->
+            not_found;
+        {ok, [LastKey | _]} ->
+            maybe_remove_messages_prefix(LastKey)
+    end.
+
+item_path(Key) ->
+    gen_server:call(?MODULE, {item_path, Key}, ?TIMEOUT).
+
+%% Helper function to resolve a chain of IDs.
+resolve_chain(Opts, BaseId, ResolutionKeys) ->
+    lists:foldl(fun (_Key, not_found) ->
+                        not_found;
+                    (Key, Acc) ->
+                        case read(Opts, Acc) of
+                            not_found ->
+                                not_found;
+                            {ok, ItemBinary} ->
+                                extract_next_id(ItemBinary, Key)
+                        end
+                end,
+                BaseId,
+                ResolutionKeys).
+
+%% Helper function to resolve the next ID from an item.
+extract_next_id(ItemBinary, ResolutionKey) ->
+    Item = ar_bundles:deserialize(ItemBinary),
+    Manifest =
+        ar_bundles:parse_manifest(
+            maps:get(<<"manifest">>, Item#tx.data)),
+    NextId = maps:get(list_to_binary(ResolutionKey), Manifest),
+    <<"messages-", NextId/binary>>.
+
+%% Fallback resolution if primary resolution fails.
+fallback_resolve(Opts, BaseId, ResolutionKeys) ->
+    Path = path(Opts, ResolutionKeys),
+    case item_path(<<BaseId/binary, "-", Path/binary>>) of
+        not_found ->
+            not_found;
+        {ok, [Key | _]} ->
+            maybe_remove_messages_prefix(Key)
+    end.
+
+-spec type(Opts, Key) -> Result
+    when Opts :: map(),
+         Key :: binary(),
+         Result :: composite | simple | not_found.
+type(Opts, ["messages", Path]) ->
+    MessagesPath = <<"messages-", Path/binary>>,
+    type(Opts, MessagesPath);
+type(Opts, Key) when is_binary(Key) ->
+    % The fs adapter looks on the FS. If it's a directory it says 'comosite'
+    % composite item is an item which has manifest!
+    case read(Opts, Key) of
+        not_found ->
+            not_found;
+        {ok, ItemBinary} ->
+            Item = ar_bundles:deserialize(ItemBinary),
+            case ar_bundles:manifest(Item) of
+                undefined ->
+                    simple;
+                _Manifest ->
+                    composite
+            end
+    end.
+
+make_group(_Opts, _Path) ->
+    % FS just creates a folder....
+    % when we write a composite item
+    % (the item which consists of multiple items) we do:
+    % 1. create folder
+    % 2. create manifest file
+    % 3. create items themselves
+    ok.
+
+-spec make_link(any(), key(), key()) -> ok.
+make_link(_, Key1, Key1) ->
+    ok;
+make_link(Opts, Key1, Key2) when is_list(Key1), is_list(Key2) ->
+    Key1Bin = list_to_binary(Key1),
+    Key2Bin = list_to_binary(Key2),
+    make_link(Opts, Key1Bin, Key2Bin);
+make_link(Opts, Key1, Key2) ->
+    ok = write(Opts, Key2, <<?LINK_PREFIX/binary, Key1/binary>>).
```

Comment:
This is an interesting approach. We cannot easily have a column with the type of the data or similar? What happens if the stored data actually starts with the LINK_PREFIX in this implementation?

If we cannot easily add a column for this another approach could be to use a type tag of specified size, such that we know for sure how to interpret it.

Reply:
@samcamwilliams That's the approach I want to discuss and validate with you:

I had to make two such decisions to replicate the behavior of the ao_fs_storage and implement ao_storage.

### storing links as "link-<my_linked_key>
This technical decision was driven by the requirement of extracting linked items by the key. The ao_fs_store implements it in the following way:
```
read(Opts = #{ dir := DataDir }, Key) ->
    case read(join([DataDir, Key])) of
        not_found ->
            case resolve(Opts, Key) of
                Key -> not_found;
                ResolvedPath -> read(ResolvedPath)
            end;
        Result -> Result
    end.
read(Path) ->
    ?c({read, Path}),
    case file:read_file_info(Path) of
        {ok, #file_info{type = regular}} ->
            {ok, _} = file:read_file(Path);
        _ ->
            case file:read_link(Path) of
                {ok, Link} ->
                    read(Link);
                _ ->
                    not_found
            end
    end.
```

I understand this way: 
1. We take the key and identify the file on the filesystem
2. We read the file's metadata to see if it's regular of a symlink
3. We read the file or the file(s) under the symlink.

The Rocksdb, on the other hand, is a key value storage that does not allow storing meta information like symlinks between items. It expects both key and value to be binaries. So, I decided to use the Value to encode the link.

I agree that this way, we may (and probably will) introduce the issue when the data collides and ground the implementation on the fact that we seem to store serialized TX records as the data otherwise. 

Regarding possible solutions:

**Colum families (tables)**
Rocksdb does not introduce tables. However, it's possible to make column families that separate different data types. For example, potentially, we can have a table(family) for links and a table(family) for items. However, it would require us to make two requests if we would want to replicate the same behavior as on the filesystem. For example
1. We read the item by Key and get some Value there (that might be short key or long serialized binary)
2. We try to query the links table to see if the item with a given key can be found here
3. We try to follow the link if found

**Encoding item types**
Another solution might be to prefix the data (every item) with a prefix of the given length so, for example
 -  <<data
 - <<link


### storing composite items as "<mykey>-item"
That's another assumption I have made.

The fs_store does the following for storing composite item (Key, Value):
1. Create a group (folder) under the given path called Key
2. Creates **item** file there, where we store Item with Manifest
3. Creates Subitems defined in manifest under the messages folder (aka main folder)
4. Create links with the folder to items created on step3

What happens in the **ao_rocksdb_store** is
1. The path from ao_cache is resolved to **Key-item**, so we store the manifest item under the (<<Key/binary, <<"-item">>, Value)
2. Individual items are created as usual (SubKey, Value)
3. Link items are created;

I would be happy to discuss them in person or here if possible.

Reply:
@oltarasenko , I don't think Prefix Mapping for Namespace Emulation will work, we will have billions of keys and the odds that a key matches the pre-fix is very likely. Even if it is a 2 step process, the best approach that I can see is the Column Families approach having 1 Family for `Links` and 1 Family for `Items`. 

I am not very familiar with RocksDB, but if the symlink_key and the target_key could be the same key, I think you will need to leverage Column Families to separate.

Happy to talk through this with you